### PR TITLE
CRUD with BlackDuck Helm Chart and Blackduck Migration from Operator to Helm

### PR DESCRIPTION
**Pull request by jdartigalongue**

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/1293#" title="Author self-approved">jdartigalongue</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/develop/OWNERS)~~ [jdartigalongue]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/synopsysctl/cmd_get.go [hunk 1]
```diff
@@ -106,15 +105,19 @@ var getBlackDuckCmd = &cobra.Command{
...
+		}
 		return nil
 	},
 	RunE: func(cmd *cobra.Command, args []string) error {
-		log.Debugf("getting Black Duck instances...")
-		out, err := RunKubeCmd(restconfig, kubeClient, generateKubectlGetCommand("blackducks", args)...)
+		helmRelease, err := util.GetWithHelm3(args[0], namespace, kubeConfigPath)
 		if err != nil {
-			return fmt.Errorf("error getting Black Duck instances due to %+v - %s", out, err)
+			return fmt.Errorf("failed to get Polaris-Reporting values: %+v", err)
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_get.go]: This should say Black Duck instead of Polaris-Reporting

File: pkg/util/helm_helpers.go [hunk 2]
```diff
@@ -347,6 +363,29 @@ func ReleaseExists(releaseName, namespace, kubeConfig string) bool {
...
+	for _, fileName := range extraFiles {
+		found := false
+		for _, chartFile := range ch.Files {
+			if fileName == chartFile.Name {
+				found = true
+				patch := make(map[string]interface{})
+				if err := yaml.Unmarshal(chartFile.Data, &patch); err != nil {
+					return err
+				}
+				if err := mergo.Merge(&vals, patch, mergo.WithOverride); err != nil {
```
Comment (mphammer) [inline pkg/util/helm_helpers.go]: I'm not sure why the Helm team wrote their own function to do this merging: https://github.com/helm/helm/blob/9b42702a4bced339ff424a78ad68dd6be6e1a80a/pkg/cli/values/options.go#L88

Comment (mphammer) [reply pkg/util/helm_helpers.go]: fyi this is how they are doing their merging: https://github.com/helm/helm/blob/9b42702a4bced339ff424a78ad68dd6be6e1a80a/pkg/cli/values/options.go#L45    :)

File: pkg/synopsysctl/constants.go [hunk 3]
```diff
@@ -53,6 +53,7 @@ var nativeClusterType = clusterTypeKubernetes
 
 var baseURL = "https://raw.githubusercontent.com/blackducksoftware/releases/master"
 
+var blackduckChartRepository = "https://artifactory.internal.synopsys.com/artifactory/bds-hub-helm-snapshot-local/blackduck/blackduck-2020.4.0.tgz"
```
Comment (mphammer) [inline pkg/synopsysctl/constants.go]: You could split this into a baseBlackDuckRepository variable so that we don't have to edit each function if it changes
This could be helpful if we break --chart-location-path into separate flags for --chart-name --chart-repo --chart-version

File: pkg/synopsysctl/cmd_create.go [hunk 4]
```diff
@@ -257,74 +246,19 @@ func checkSealKey(flagset *pflag.FlagSet) {
...
-	if !cmd.Flags().Lookup("template").Changed {
-		baseBlackDuckSpec = defaultBaseBlackDuckSpec
-	}
-	log.Debugf("setting Black Duck's base spec to '%s'", baseBlackDuckSpec)
-	err := createBlackDuckCobraHelper.SetPredefinedCRSpec(baseBlackDuckSpec)
-	if err != nil {
-		cmd.Help()
-		return err
-	}
+	//// Set the base spec
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: We can probably delete this function or delete the commented out code

File: pkg/synopsysctl/cmd_create.go [hunk 5]
```diff
@@ -257,74 +246,19 @@ func checkSealKey(flagset *pflag.FlagSet) {
...
-		return nil, fmt.Errorf("failed to get Default App: %+v", err)
-	}
-	defaultPvcComponentsList, err := app.Blackduck().GetComponents(&blackduckv1.Blackduck{Spec: bd.Spec}, blackduckapp.PVCResources)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get PVC Components List: %+v", err)
-	}
-	return defaultPvcComponentsList.PersistentVolumeClaims, nil
-}
-
 func convertHorizonPVCComponentToBlackDuckPVC(horizonPVCList []*components.PersistentVolumeClaim) []blackduckv1.PVC {
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: I think this function can be deleted too

File: pkg/synopsysctl/cmd_create.go [hunk 6]
```diff
@@ -358,103 +292,58 @@ var createBlackDuckCmd = &cobra.Command{
...
-				return fmt.Errorf("error setting Spec with PVC values: %s", err)
-			}
-
-			// Update the CR in createBlackDuckCobraHelper with user's flags
-			cmd.Flag("pvc-file-path").Changed = false // we already did the special logic above to set the PVCs
-			blackDuck, err := updateBlackDuckSpecWithFlags(cmd, blackDuckName, blackDuckNamespace)
-			if err != nil {
-				return err
-			}
+		fmt.Printf("Helm Map\n%+v\n", helmValuesMap)
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: Do you want this print statement?

File: pkg/synopsysctl/cmd_create.go [hunk 7]
```diff
@@ -917,7 +794,7 @@ var createPolarisReportingCmd = &cobra.Command{
 		} else {
 			versionFlag := cmd.Flag("version")
 			if versionFlag.Changed {
-				polarisReportingChartRepository = fmt.Sprintf("%s/charts/polaris-helmchart-reporting-%s.tgz", baseChartRepository, versionFlag.Value.String())
+				polarisReportingChartRepository = fmt.Sprintf("https://chartmuseum.polaris-cc-staging.sig-clops.synopsys.com/charts/polaris-helmchart-reporting-%s.tgz", versionFlag.Value.String())
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: Why remove baseChartRepository?

File: pkg/synopsysctl/cmd_create.go [hunk 8]
```diff
@@ -358,103 +292,58 @@ var createBlackDuckCmd = &cobra.Command{
...
-		versionSupportsSecurityContexts, err := util.IsVersionGreaterThanOrEqualTo(blackDuck.Spec.Version, 2019, time.December, 0)
-		if err != nil {
-			return fmt.Errorf("failed to check Black Duck version: %s", err)
-		}
-		if !versionSupportsSecurityContexts && cmd.Flags().Changed("security-context-file-path") {
-			return fmt.Errorf("security contexts from --security-context-file-path cannot be set for versions before 2019.12.0, you're using version %s", blackDuck.Spec.Version)
-		}
-		if util.IsOpenshift(kubeClient) && cmd.Flags().Changed("security-context-file-path") {
-			return fmt.Errorf("cannot set security contexts with --security-context-file-path in an Openshift environment")
+		for _, v := range secrets {
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: For Polaris-Reporting we used `KubectlApplyRuntimeObjects` so that it doesn't care about the type of resource. I think this is okay tho :) and it helps us move away from being dependent on the kubectl binary. I'm going to create a ticket to make a function that creates any resource with the KubeClient

File: pkg/synopsysctl/cmd_delete.go [hunk 9]
```diff
@@ -82,29 +82,21 @@ var deleteAlertCmd = &cobra.Command{
 // deleteBlackDuckCmd deletes Black Duck instances from the cluster
 var deleteBlackDuckCmd = &cobra.Command{
 	Use:           "blackduck NAME...",
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_delete.go]: Can you please remove the ... since it only does 1 instance now?

File: pkg/synopsysctl/cmd_start.go [hunk 10]
```diff
@@ -96,42 +96,24 @@ var startBlackDuckCmd = &cobra.Command{
...
-			}
-			log.Infof("starting Black Duck '%s' in namespace '%s'...", blackDuckName, blackDuckNamespace)
-
-			// Get the Black Duck
-			currBlackDuck, err := util.GetBlackduck(blackDuckClient, crdNamespace, blackDuckName, metav1.GetOptions{})
-			if err != nil {
-				errors = append(errors, fmt.Errorf("error getting Black Duck '%s' in namespace '%s' due to %+v", blackDuckName, blackDuckNamespace, err))
-				continue
-			}
+		chartRepo := "https://artifactory.internal.synopsys.com/artifactory/bds-hub-helm-snapshot-local/blackduck/blackduck-2020.4.0.tgz"
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_start.go]: I think we need to change this to handle if the user provides --chart-location-path or a different --version

File: pkg/synopsysctl/cmd_stop.go [hunk 11]
```diff
@@ -102,45 +102,25 @@ var stopBlackDuckCmd = &cobra.Command{
...
-			blackDuckNamespace, crdNamespace, scope, err := getInstanceInfo(false, util.BlackDuckCRDName, util.BlackDuckName, namespace, blackDuckName)
-			if err != nil {
-				if len(blackDuckNamespace) == 0 && scope == apiextensions.ClusterScoped {
-					err = fmt.Errorf("%s %s doesn't appear to be running: %v", util.BlackDuckName, blackDuckName, err)
-				}
-				errors = append(errors, err)
-				continue
-			}
-			log.Infof("stopping Black Duck '%s' in namespace '%s'...", blackDuckName, blackDuckNamespace)
+		chartRepo := "https://artifactory.internal.synopsys.com/artifactory/bds-hub-helm-snapshot-local/blackduck/blackduck-2020.4.0.tgz"
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_stop.go]: I think we need to handle --chart-location-path and --version here

File: pkg/blackduck/ctl_blackduck.go [hunk 12]
```diff
@@ -120,39 +112,10 @@ const (
...
-		return fmt.Errorf("Black Duck spec type '%s' is not valid", specType)
-	}
-	return nil
-}
-
 // AddCRSpecFlagsToCommand adds flags to a Cobra Command that are need for BlackDuck's Spec.
 // The flags map to fields in the CRSpecBuilderFromCobraFlags struct.
 // master - if false, doesn't add flags that all Users shouldn't use
-func (ctl *CRSpecBuilderFromCobraFlags) AddCRSpecFlagsToCommand(cmd *cobra.Command, master bool) {
+func (ctl *HelmValuesFromCobraFlags) AddCRSpecFlagsToCommand(cmd *cobra.Command, master bool) {
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: Can we remove the DesiredState flag? It has been deprecated for a while now: https://github.com/blackducksoftware/synopsys-operator/pull/1293/files#diff-2ebafce1624553553e121ddbe1c87469R165

File: pkg/blackduck/ctl_blackduck.go [hunk 13]
```diff
@@ -265,161 +220,72 @@ func FlagWasSet(flagset *pflag.FlagSet, flagName string) bool {
...
-		ctl.blackDuckSpec.Environs = finalEnvironValues
-	}
-	return nil
-}
-
 // SetCRSpecFieldByFlag updates a field in the blackDuckSpec if the flag was set by the user. It gets the
 // value from the corresponding struct field.
 // Note: It should only handle values with a 1 to 1 mapping - struct-field to spec
-func (ctl *CRSpecBuilderFromCobraFlags) SetCRSpecFieldByFlag(f *pflag.Flag) {
+func (ctl *HelmValuesFromCobraFlags) AddHelmValueByCobraFlag(f *pflag.Flag) {
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: The comment has the wrong function name now ^

File: pkg/blackduck/ctl_blackduck.go [hunk 14]
```diff
@@ -265,161 +220,72 @@ func FlagWasSet(flagset *pflag.FlagSet, flagName string) bool {
...
-		case "version":
-			ctl.blackDuckSpec.Version = ctl.Version
-		case "expose-ui":
-			ctl.blackDuckSpec.ExposeService = ctl.ExposeService
-		case "db-prototype":
-			ctl.blackDuckSpec.DbPrototype = ctl.DbPrototype
-		case "external-postgres-host":
-			if ctl.blackDuckSpec.ExternalPostgres == nil {
-				ctl.blackDuckSpec.ExternalPostgres = &blackduckv1.PostgresExternalDBConfig{}
+			util.SetHelmValueInMap(ctl.args, []string{"size"}, ctl.Size)
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: The BlackDuck Values.yaml doesn't have a `size` field

File: pkg/blackduck/ctl_blackduck.go [hunk 15]
```diff
@@ -443,71 +313,32 @@ func (ctl *CRSpecBuilderFromCobraFlags) SetCRSpecFieldByFlag(f *pflag.Flag) {
...
+			util.SetHelmValueInMap(ctl.args, []string{"status"}, ctl.DesiredState)
 		case "migration-mode":
-			if ctl.MigrationMode {
-				ctl.blackDuckSpec.DesiredState = "DbMigrate"
-			}
-		case "image-registries":
-			ctl.blackDuckSpec.ImageRegistries = ctl.ImageRegistries
-		case "license-key":
-			ctl.blackDuckSpec.LicenseKey = ctl.LicenseKey
+			util.SetHelmValueInMap(ctl.args, []string{"status"}, "DbMigrate")
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: I don't think the BlackDuck Helm Chart can handle DbMigrate right now

File: pkg/blackduck/ctl_blackduck.go [hunk 16]
```diff
@@ -85,25 +80,22 @@ type CRSpecBuilderFromCobraFlags struct {
 }
 
 // NewCRSpecBuilderFromCobraFlags creates a new CRSpecBuilderFromCobraFlags type
-func NewCRSpecBuilderFromCobraFlags() *CRSpecBuilderFromCobraFlags {
-	return &CRSpecBuilderFromCobraFlags{
-		blackDuckSpec: &blackduckv1.BlackduckSpec{},
+func NewHelmValuesFromCobraFlags() *HelmValuesFromCobraFlags {
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: Function name is wrong in the comments

File: pkg/blackduck/ctl_blackduck.go (hunk 14 above)
Comment (jdartigalongue) [reply pkg/blackduck/ctl_blackduck.go]: Correct, since our Helm chart doesn't support sizes directly from the configuration file we need to have this hack in order to store the size somewhere. Synopsysctl is then responsible for reading the "size" value from the config and apply the patch.

Comment (mphammer) [reply pkg/blackduck/ctl_blackduck.go]: Oh gotcha! That's an interesting hack. Synopsysctl could use check the cmd.Flags("size").Value.ToString() and cmd.Flags("size").changed, but this works too :)

Comment (jdartigalongue) [reply pkg/blackduck/ctl_blackduck.go]: The problem with only relying on the flag is that a user would have to specify the size every single time they update since the size wouldn't be stored anywhere

File: pkg/synopsysctl/cmd_create.go (hunk 7 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/cmd_create.go]: Small issue during a rebase :D

File: pkg/blackduck/ctl_blackduck.go (hunk 15 above)
Comment (jdartigalongue) [reply pkg/blackduck/ctl_blackduck.go]: @msenmurugan Is there any plan to add this feature or should we just get rid of this?

File: pkg/blackduck/ctl_blackduck.go (hunk 14 above)
Comment (mphammer) [reply pkg/blackduck/ctl_blackduck.go]: Oh that's true! Okay I'm on board :)

File: pkg/blackduck/ctl_blackduck.go [hunk 17]
```diff
@@ -443,71 +311,29 @@ func (ctl *CRSpecBuilderFromCobraFlags) SetCRSpecFieldByFlag(f *pflag.Flag) {
...
-				ctl.blackDuckSpec.RegistryConfiguration = &api.RegistryConfiguration{}
-			}
-			ctl.blackDuckSpec.RegistryConfiguration.Registry = ctl.Registry
+			util.SetHelmValueInMap(ctl.args, []string{"registry"}, ctl.Registry)
 		case "pull-secret-name":
-			if ctl.blackDuckSpec.RegistryConfiguration == nil {
-				ctl.blackDuckSpec.RegistryConfiguration = &api.RegistryConfiguration{}
-			}
-			ctl.blackDuckSpec.RegistryConfiguration.PullSecrets = ctl.PullSecrets
+			util.SetHelmValueInMap(ctl.args, []string{"imagePullSecrets"}, ctl.PullSecrets)
```
Comment (mphammer) [inline pkg/blackduck/ctl_blackduck.go]: So I think we are going to have to change this to create a list of maps
```
 imagePullSecrets
 - name: registryName
```
Maybe like this...
```
pullSecretList := []map[string]interface{}{}
for loop ctl.imagePullSecrets {
    key, val := imagePullSecret.Split(":") // How are imagePullSecrets formatted?
    pullSecretList = append(pullSecretList, map[string]interface{}{key: val}
}
util.SetHelmValueInMap(ctl.args, []string{"imagePullSecrets"}, pullSecretList)
```

File: pkg/synopsysctl/cmd_get.go [hunk 18]
```diff
@@ -100,21 +99,25 @@ var getAlertCmd = &cobra.Command{
 // getBlackDuckCmd display one or many Black Duck instances
 var getBlackDuckCmd = &cobra.Command{
 	Use:           "blackduck [NAME...]",
-	Example:       "synopsysctl get blackducks\nsynopsysctl get blackduck <name>\nsynopsysctl get blackducks <name1> <name2>\nsynopsysctl get blackducks -n <namespace>\nsynopsysctl get blackduck <name> -n <namespace>\nsynopsysctl get blackducks <name1> <name2> -n <namespace>",
+	Example:       "ynopsysctl get blackduck <name> -n <namespace>",
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_get.go]: missing "s" in ynopsysctl

File: pkg/synopsysctl/cmd_create.go [hunk 19]
```diff
@@ -984,7 +820,7 @@ var createPolarisReportingNativeCmd = &cobra.Command{
 		} else {
 			versionFlag := cmd.Flag("version")
 			if versionFlag.Changed {
-				polarisReportingChartRepository = fmt.Sprintf("%s/charts/polaris-helmchart-reporting-%s.tgz", baseChartRepository, versionFlag.Value.String())
+				polarisReportingChartRepository = fmt.Sprintf("https://chartmuseum.polaris-cc-staging.sig-clops.synopsys.com/charts/polaris-helmchart-reporting-%s.tgz", versionFlag.Value.String())
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_create.go]: Was this going to be changed back to use the baseChartRepository?

File: pkg/synopsysctl/cmd_stop.go [hunk 20]
```diff
@@ -102,45 +102,36 @@ var stopBlackDuckCmd = &cobra.Command{
...
-			log.Infof("successfully submitted stop Black Duck '%s' in namespace '%s'", blackDuckName, blackDuckNamespace)
+		helmValuesMap, err := createBlackDuckCobraHelper.GenerateHelmFlagsFromCobraFlags(cmd.Flags())
+		if err != nil {
+			return err
 		}
-		if len(errors) > 0 {
-			return fmt.Errorf("%v", errors)
+		util.SetHelmValueInMap(helmValuesMap, []string{"status"}, "Stopped")
+
+		// Deploy Polaris-Reporting Resources
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_stop.go]: This should say Black Duck Resources

File: pkg/synopsysctl/cmd_db_migrate.go [hunk 21]
```diff
@@ -48,42 +45,30 @@ var dbMigrateBlackDuckCmd = &cobra.Command{
...
-		if len(errors) > 0 {
-			return fmt.Errorf("%v", errors)
+		util.SetHelmValueInMap(helmValuesMap, []string{"status"}, "dbmigrate")
+		// Check Dry Run before deploying any resources
+		err = util.CreateWithHelm3(args[0], namespace, chartRepo, helmValuesMap, kubeConfigPath, true)
+		if err != nil {
+			return fmt.Errorf("failed to create Blackduck resources: %+v", err)
+		}
+
+		// Deploy Polaris-Reporting Resources
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_db_migrate.go]: This is just the Create command right? Is it doing the database migrate? 
I think this might depend on if we modify the Helm Chart to have "dbMigrate" as an option for .Values.status

File: pkg/synopsysctl/blackduck_migrate.go [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+	}
+
+	log.Info("Deleting Blackduck resources")
+	// TODO wait for resources to be deleted
+	if err := deleteComponents(bd); err != nil {
+		return err
+	}
+
+	log.Info("Upgrading Blackduck using Helm")
+	chartRepo := "https://artifactory.internal.synopsys.com/artifactory/bds-hub-helm-snapshot-local/blackduck/blackduck-2020.4.0.tgz"
```
Comment (mphammer) [inline pkg/synopsysctl/blackduck_migrate.go]: You might already be planning to fix all the chartRepo variables but just in case this is hard coded here :)

File: pkg/synopsysctl/cmd_db_migrate.go (hunk 21 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/cmd_db_migrate.go]: Let's ignore db migrate until we decide what to do. If we need to maintain this feature then it first needs to be added to the chart.

File: pkg/blackduck/ctl_blackduck.go (hunk 17 above)
Comment (jdartigalongue) [reply pkg/blackduck/ctl_blackduck.go]: done

File: pkg/synopsysctl/cmd_db_migrate.go (hunk 21 above)
Comment (mphammer) [reply pkg/synopsysctl/cmd_db_migrate.go]: Sounds good to me

File: pkg/blackduck/ctl_blackduck.go (hunk 17 above)
Comment (mphammer) [reply pkg/blackduck/ctl_blackduck.go]: Ah cool nice!

File: pkg/synopsysctl/cmd_delete.go [hunk 23]
```diff
@@ -75,30 +75,22 @@ var deleteAlertCmd = &cobra.Command{
...
-			if err != nil {
-				return err
-			}
-			log.Infof("deleting Black Duck '%s' in namespace '%s'...", blackDuckName, blackDuckNamespace)
-			err = util.DeleteBlackduck(blackDuckClient, blackDuckName, crdNamespace, &metav1.DeleteOptions{})
-			if err != nil {
-				return fmt.Errorf("error deleting Black Duck '%s' in namespace '%s' due to '%s'", blackDuckName, blackDuckNamespace, err)
-			}
-			log.Infof("successfully submitted delete Black Duck '%s' in namespace '%s'", blackDuckName, blackDuckNamespace)
+		err := util.DeleteWithHelm3(args[0], namespace, kubeConfigPath)
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_delete.go]: Do we need to delete the certificate secrets here?

File: pkg/synopsysctl/cmd_start.go [hunk 24]
```diff
@@ -91,47 +91,41 @@ var startAlertCmd = &cobra.Command{
...
+		// Update the Helm Chart Location
+		chartLocationFlag := cmd.Flag("chart-location-path")
+		if chartLocationFlag.Changed {
+			blackduckChartRepository = chartLocationFlag.Value.String()
+		} else {
+			blackduckChartRepository = fmt.Sprintf("%s/charts/blackduck-%s.tgz", baseChartRepository, instance.Config["imageTag"])
+		}
 
-			log.Infof("successfully submitted start Black Duck '%s' in namespace '%s'", blackDuckName, blackDuckNamespace)
+		helmValuesMap, err := createBlackDuckCobraHelper.GenerateHelmFlagsFromCobraFlags(cmd.Flags())
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_start.go]: You might get that null pointer error again here

File: pkg/synopsysctl/cmd_start.go [hunk 25]
```diff
@@ -91,47 +91,38 @@ var startAlertCmd = &cobra.Command{
...
+			return fmt.Errorf("couldn't find instance %s in namespace %s", args[0], namespace)
+		}
 
-			log.Infof("successfully submitted start Black Duck '%s' in namespace '%s'", blackDuckName, blackDuckNamespace)
+		// Update the Helm Chart Location
+		chartLocationFlag := cmd.Flag("chart-location-path")
+		if chartLocationFlag.Changed {
+			blackduckChartRepository = chartLocationFlag.Value.String()
+		} else {
+			blackduckChartRepository = fmt.Sprintf("%s/charts/blackduck-%s.tgz", baseChartRepository, instance.Config["imageTag"])
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_start.go]: I think instance.Config only has the values that were set with flags (not all values in Values.yaml). So if the user didn't specify --version then this might be empty.

Comment (mphammer) [reply pkg/synopsysctl/cmd_start.go]: I think we can use `instance.Chart.Values["imageTag"]`

File: pkg/synopsysctl/cmd_update.go [hunk 26]
```diff
@@ -486,150 +186,100 @@ var updateBlackDuckCmd = &cobra.Command{
...
-				r, _ := regexp.Compile("blackduck-.*")
-				pvcNameKey := r.FindString(pvc.Name) // removes the "<blackduckName>-" from the PvcName
-				sc := appsutil.GetSecurityContext(currBlackDuck.Spec.SecurityContexts, pvcNameToSecurityContextNameMap[pvcNameKey])
-				if sc != nil {
-					if sc.RunAsUser != nil {
-						pvcNameToFileOwnershipMap[pvc.Name] = *sc.RunAsUser
-					}
-				}
+		} else if isOperatorBased {
+			ok, err := util.IsVersionGreaterThanOrEqualTo(cmd.Flag("version").Value.String(), 2019, time.April, 0)
```
Comment (mphammer) [inline pkg/synopsysctl/cmd_update.go]: What happens if cmd.Flag("version") is not passed into the update command? 
Theoretically this code should only be executed if the user is "upgrading" an old BlackDuck version to a new one so it must have been set. But maybe we should put a check in for if a customer accidentally tries to "update" a different field of an old BlackDuck.

Comment (mphammer) [reply pkg/synopsysctl/cmd_update.go]: This is what I put in Alert:
```
if !isOperatorBased && instance != nil {
	err = updateAlertHelmBased(cmd, alertName)
} else if isOperatorBased {
	// user must be upgrading to a new version
	if !cmd.Flag("version").Changed { // TODO fill in the alert version
		return fmt.Errorf("you must upgrade this Alert version with --version to use this synopsysctl binary - modifying Alert versions before XXXXX are not supported with this binary")
	}
	err = updateAlertOperatorBased()
}
```

File: cmd/operator-ui/actions/blackducks.go [hunk 27]
```diff
@@ -117,7 +117,7 @@ func (v BlackducksResource) New(c buffalo.Context) error {
 	blackduck := &blackduckapi.Blackduck{}
 	blackduck.Spec = *blackduckSpec
 	// TODO: fix the sort logic for Black Duck version
-	blackduck.Spec.Version = "2020.2.1"
+	blackduck.Spec.Version = "2020.2.0"
```
Comment (yashbhutwala) [inline cmd/operator-ui/actions/blackducks.go]: do we need a separate PR for getting rid of the UI?  are we tracking it somewhere?

File: pkg/blackduck/certificate.go [hunk 28]
```diff
@@ -116,3 +120,78 @@ func CreateSelfSignedCert() (string, string) {
 	pem.Encode(key, pemBlockForKey(priv))
 	return certificate.String(), key.String()
 }
+
+func GetCertificateSecretFromFile(secretName, namespace, certPath, keyPath string) (*corev1.Secret, error) {
```
Comment (yashbhutwala) [inline pkg/blackduck/certificate.go]: can we move this into the helm chart itself? ... maybe not in this PR, but if we're ok with that, I'll make a ticket to track it

File: pkg/blackduck/ctl_blackduck.go (hunk 15 above)
Comment (yashbhutwala) [reply pkg/blackduck/ctl_blackduck.go]: there is already a stop status, which should be enough.

File: pkg/synopsysctl/blackduck_migrate.go [hunk 29]
```diff
@@ -0,0 +1,370 @@
...
+			return false
+		}
+	}
+	return false
+}
+
+func BlackduckV1ToHelm(bd *v1.Blackduck, operatorNamespace string) (map[string]interface{}, error) {
+	helmConfig := make(map[string]interface{})
+
+	// TODO handle sizes
```
Comment (yashbhutwala) [inline pkg/synopsysctl/blackduck_migrate.go]: can you bubble up the leftover TODOs in the description in this PR?  or some ticket?

File: pkg/synopsysctl/blackduck_migrate.go [hunk 30]
```diff
@@ -0,0 +1,370 @@
...
+
+	if isFeatureEnabled(bd.Spec.Environs, "ENABLE_SOURCE_UPLOADS", "true") {
+		util.SetHelmValueInMap(helmConfig, []string{"enableSourceCodeUpload"}, true)
+	}
+
+	if isFeatureEnabled(bd.Spec.Environs, "USE_BINARY_UPLOADS", "1") {
+		util.SetHelmValueInMap(helmConfig, []string{"enableBinaryScanner"}, true)
+	}
+
+	// TODO
```
Comment (yashbhutwala) [inline pkg/synopsysctl/blackduck_migrate.go]: this one is already done based on what I see

File: pkg/synopsysctl/cmd_create.go [hunk 31]
```diff
@@ -354,103 +252,55 @@ var createBlackDuckCmd = &cobra.Command{
 			return fmt.Errorf("this command takes 1 argument")
 		}
 		checkPasswords(cmd.Flags())
+		cobra.MarkFlagRequired(cmd.Flags(), "certificate-file-path")
+		cobra.MarkFlagRequired(cmd.Flags(), "certificate-key-file-path")
```
Comment (yashbhutwala) [inline pkg/synopsysctl/cmd_create.go]: interesting these were not required before

File: pkg/synopsysctl/cmd_create_test.go [hunk 32]
```diff
@@ -1,135 +0,0 @@
...
-	//blackduckv1 "github.com/blackducksoftware/synopsys-operator/pkg/api/blackduck/v1"
-	"reflect"
-
-	// "github.com/blackducksoftware/synopsys-operator/pkg/blackduck"
-	"github.com/blackducksoftware/synopsys-operator/pkg/util"
-	"github.com/spf13/cobra"
-)
-
-func TestAddPVCValuesToBlackDuckSpec(t *testing.T) {
-
```
Comment (yashbhutwala) [inline pkg/synopsysctl/cmd_create_test.go]: maybe we should replace these with their according unit tests? 😃

File: pkg/synopsysctl/blackduck_migrate.go (hunk 29 above)
Comment (mphammer) [reply pkg/synopsysctl/blackduck_migrate.go:129]: Tickets would be best! I have found that people miss it otherwise

File: pkg/synopsysctl/cmd_delete.go (hunk 23 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/cmd_delete.go:92]: Good catch! It's been updated.

File: pkg/synopsysctl/cmd_start.go (hunk 25 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/cmd_start.go]: Done

File: pkg/synopsysctl/cmd_update.go (hunk 26 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/cmd_update.go:242]: Good catch!

File: cmd/operator-ui/actions/blackducks.go (hunk 27 above)
Comment (jdartigalongue) [reply cmd/operator-ui/actions/blackducks.go]: The UI has been removed

File: pkg/synopsysctl/blackduck_migrate.go (hunk 29 above)
Comment (jdartigalongue) [reply pkg/synopsysctl/blackduck_migrate.go]: The size is now handled during the migration

File: pkg/apps/alert/latest/alert_versions.go [hunk 33]
```diff
@@ -77,10 +73,6 @@ var imageTags = map[string]map[string]string{
 		"blackduck-alert": "5.2.3",
 		"blackduck-cfssl": "1.0.0",
 	},
```
Comment (mphammer) [inline pkg/apps/alert/latest/alert_versions.go:75]: Why are these being removed?

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/1293#" title="Author self-approved">jdartigalongue</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/1293#pullrequestreview-386748757" title="LGTM">mphammer</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/develop/OWNERS)~~ [jdartigalongue,mphammer]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
